akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala

```diff
@@ -0,0 +1,317 @@
+package akka.typed.scaladsl.persistence
+
+import akka.typed
+
+import scala.concurrent.duration._
+import akka.typed.{ ActorRef, Behavior, ExtensibleBehavior, Signal, Terminated }
+import akka.typed.scaladsl.{ ActorContext, TimerScheduler }
+
+import scala.concurrent.ExecutionContext
+import scala.reflect.ClassTag
+
+class ApiTest {
+  object TypedPersistentActor {
+
+    sealed trait PersistentEffect[+Event]
+
+    case object PersistNothing extends PersistentEffect[Nothing]
+
+    case class PersistAnd[Event](event: Event, callback: Event ⇒ Unit) extends PersistentEffect[Event]
+
+    case class Persist[Event](event: Event) extends PersistentEffect[Event] {
+      def andThen(callback: Event ⇒ Unit) = PersistAnd(event, callback)
+    }
+
+    class ActionHandler[Command: ClassTag, Event, State](val handler: ((ActorContext[Command], State, Any) ⇒ PersistentEffect[Event])) {
+      def onSignal(signalHandler: (ActorContext[Command], State, Any) ⇒ PersistentEffect[Event]): ActionHandler[Command, Event, State] =
+        new ActionHandler({
+          case (ctx, state, command: Command) ⇒ handler(ctx, state, command)
```

Comment:
This implies signals cannot also be commands - which seems reasonable, would be nice to ensure this statically...